subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java

```diff
@@ -405,18 +453,36 @@ public Configuration resolutionStrategy(Closure closure) {
     }
 
     private void validateMutation(boolean lenient) {
-        boolean userAlreadyNagged = false;
-        if (getState() != State.UNRESOLVED) {
-            if (!lenient) {
-                throw new InvalidUserDataException(String.format("Cannot change %s after it has been resolved.", getDisplayName()));
-            } else {
-                userAlreadyNagged = true;
+        switch (state) {
+            case UNOBSERVED:
+                // Nothing from the configuration has been observed yet, can change anything.
+                break;
+            case OBSERVED:
+                // The configuration has been used in a resolution, and it is deprecated for
+                // build logic to change any dependencies, artifacts, exclude rules or parent
+                // configurations (non-lenient properties).
+                if (!lenient) {
+                    DeprecationLogger.nagUserOfDeprecatedBehaviour(String.format("Attempting to change %s after it has been included in dependency resolution", getDisplayName()));
```

Comment:
It would be good if this message explained the types of ways that a configuration might be 'included in resolution' (or at least gave an example: 'e.g. used as parent config').